services/QuillLMS/spec/services/classlink_integration/classroom_persister_spec.rb

```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe ClasslinkIntegration::ClassroomPersister do
+  subject do
+    described_class.new(
+      incoming_classroom:,
+      existing_provider_classrooms:,
+      classlink_instance_id: classlink_instance.id
+    )
+  end
+
+  let(:incoming_classroom) do
+    {
+      sourced_id: 'external-123',
+      name: 'Math 101',
+      visible: true
+    }
+  end
+
+  describe '#run' do
+    let(:classlink_instance) { create(:classlink_instance) }
+
+    context 'when the classroom already exists' do
+      let(:existing_classroom) { create(:classroom, name: 'Old Math Class', visible: false) }
+      let(:classlink_classroom) do
+        create(:classlink_classroom,
+          external_id: 'external-123',
+          classroom: existing_classroom,
+          provider_instance: classlink_instance)
+      end
+      let(:existing_provider_classrooms) { [classlink_classroom] }
+
+      it 'updates the attributes of the existing classroom' do
+        subject
+        expect { subject.run }.to not_change(ClasslinkClassroom, :count)
+
+        existing_classroom.reload
+        expect(existing_classroom.name).to eq('Math 101')
+        expect(existing_classroom.visible).to be true
+      end
+    end
+
+    context 'classroom does not exist' do
+      let(:existing_provider_classrooms) { [] }
+
+      it 'creates new ClasslinkClassroom and Classroom records' do
+        expect { subject.run }
+          .to change(ClasslinkClassroom, :count).by(1)
+          .and change(Classroom, :count).by(1)
+      end
+    end
+
+    context 'with multiple existing classrooms' do
+      let(:matching_classroom) { create(:classroom, name: 'Matching Class') }
+      let(:matching_classlink_classroom) do
+        create(:classlink_classroom,
+          external_id: 'external-123',
+          classroom: matching_classroom,
+          provider_instance: classlink_instance)
+      end
+
+      let(:non_matching_classroom) { create(:classroom, name: 'Non-matching Class') }
+      let(:non_matching_classlink_classroom) do
+        create(:classlink_classroom,
+          external_id: 'external-456',
+          classroom: non_matching_classroom,
+          provider_instance: classlink_instance)
+      end
+
+      let(:existing_provider_classrooms) { [matching_classlink_classroom, non_matching_classlink_classroom] }
+
+      it 'only updates the matching classroom' do
+        subject.run
+        matching_classroom.reload
+        non_matching_classroom.reload
+
+        expect(matching_classroom.visible).to be true
+
+        expect(non_matching_classroom).not_to receive(:update!)
```

Comment:
Does this expectation trigger as expected when it is declared after the `subject.run` call?  I've always thought that `expect (not) to receive` had to be declared before the code that might call it, but sometimes `expect` matchers work differently than I expect.